pkg/providers/aws/provider_postgres.go

```diff
@@ -0,0 +1,487 @@
+			msg := "failed to add finalizer to instance"
+			return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrapf(err, msg)
+		}
+	}
+
+	// info about the RDS instance to be created
+	postgresCfg, _, stratCfg, err := p.getPostgresConfig(ctx, r)
+	if err != nil {
+		msg := "failed to retrieve aws RDS cluster config for instance"
+		return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrapf(err, msg)
+	}
+
+	// create the credentials to be used by the aws resource providers, not to be used by end-user
+	provoiderCreds, err := p.CredentialManager.ReconcileProviderCredentials(ctx, r.Namespace)
+	if err != nil {
+		msg := "failed to reconcile RDS credentials"
+		return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	// create credentials secret
+	if err := p.CreateSecret(ctx, buildDefaultPostgresSecret(r)); err != nil {
+		msg := "failed to create or update postgres secret"
+		return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	// setup aws RDS instance sdk session
+	cacheSvc := createPostgresService(stratCfg, provoiderCreds)
+
+	// create the aws RDS instance
+	return p.createPostgresInstance(ctx, r, cacheSvc, postgresCfg)
+}
+
+func createPostgresService(stratCfg *StrategyConfig, providerCreds *AWSCredentials) rdsiface.RDSAPI {
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region:      aws.String(stratCfg.Region),
+		Credentials: credentials.NewStaticCredentials(providerCreds.AccessKeyID, providerCreds.SecretAccessKey, ""),
+	}))
+	return rds.New(sess)
+}
+
+func (p *AWSPostgresProvider) createPostgresInstance(ctx context.Context, cr *v1alpha1.Postgres, rdsSvc rdsiface.RDSAPI, postgresCfg *rds.CreateDBInstanceInput) (*providers.PostgresInstance, v1alpha1.StatusMessage, error) {
+	// the aws access key can sometimes still not be registered in aws on first try, so loop
+	pi, err := getPostgresInstances(rdsSvc)
+	if err != nil {
+		// return nil error so this function can be requeued
+		logrus.Info("error getting replication groups : ", err)
+		return nil, "error getting replication groups", err
+	}
+
+	// verify postgresConfig
+	if err := p.verifyPostgresCreateConfig(ctx, cr, postgresCfg); err != nil {
+		msg := "failed to verify aws postgres cluster configuration"
+		return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	credSec := &v1.Secret{}
+	if err := p.Client.Get(ctx, types.NamespacedName{Name: cr.Name + defaultCredSecSuffix, Namespace: cr.Namespace}, credSec); err != nil {
+		msg := "failed to retrieve postgres credential secret"
+		return nil, v1alpha1.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+
+	for k, p := range credSec.Data {
+		if k == defaultPostgresPasswordKey {
+			defaultAwsPostgresPassword = string(p)
+		}
+	}
+
+	// check if the cluster has already been created
+	var foundInstance *rds.DBInstance
+	for _, i := range pi {
+		if *i.DBInstanceIdentifier == *postgresCfg.DBInstanceIdentifier {
+			foundInstance = i
+			break
+		}
+	}
+	if foundInstance != nil {
```

Comment:
As the longer branch of logic is there being a found instance. We could inverse this and keep this a bit flatter

```
if foundInstance == nil {
    // handle create
    // ...
    if _, err = rdsSvc.CreateDBInstance(postgresCfg); err != nil {
        // handle create err
    }
    // handle create success
}
if foundInstance.DBInstanceStatus != "available" {
    return nil, fmt.Sprintf("rds instance creation in progress, current status is %s", foundInstance.DBInstanceStatus), nil
}
// check for update ... 
``